### PR TITLE
Allow multiple edges in DAG

### PR DESCRIPTION
**Pull request by d-michail**
Minor changes in DAG to allow multiple edges + a few tests. See #765.

@jsichi @jkinable @Toptachamann Could you take a look to make sure I have not missed anything. It seems to me that our implementation works fine even with DAGs with multiple edges.

----
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-core/src/main/java/org/jgrapht/graph/DirectedAcyclicGraph.java [hunk 1]
```diff
@@ -124,7 +123,12 @@ protected DirectedAcyclicGraph(
         VisitedStrategyFactory visitedStrategyFactory, TopoOrderMap<V> topoOrderMap,
         boolean weighted)
     {
-        super(vertexSupplier, edgeSupplier, weighted);
+        super(
+            vertexSupplier, edgeSupplier,
+            new DefaultGraphType.Builder()
+                .directed().allowMultipleEdges(true).allowSelfLoops(false).weighted(weighted)
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/DirectedAcyclicGraph.java]: Shouldn't we allow the caller to determine whether multiple edges are allowed (with the default being false for backwards compatibility)?

Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/graph/DirectedAcyclicGraph.java]: Any ideas how to do that? Should we add another constructor with a `multipleEdges` parameter? Or  should we add a constructor with a `GraphType` parameter?

Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/graph/DirectedAcyclicGraph.java]: Added backwards compatibility as you suggested.

File: jgrapht-core/src/main/java/org/jgrapht/graph/DirectedAcyclicGraph.java [hunk 2]
```diff
@@ -118,20 +133,50 @@ public DirectedAcyclicGraph(
      * @param topoOrderMap the topological order map. For performance reasons, subclasses can change
      *        the way this class stores the topological order.
      * @param weighted if true the graph will be weighted, otherwise not
+     * @deprecated Deprecated
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/graph/DirectedAcyclicGraph.java]: Why deprecated?

Comment (d-michail) [reply jgrapht-core/src/main/java/org/jgrapht/graph/DirectedAcyclicGraph.java]: Took it back! Mostly to avoid maintaining two protected constructors.

**Comment by jsichi** [comment]
Looks good, thanks!

**Pull request merged**